src/rust/engine/options/src/args.rs

```diff
@@ -3,95 +3,142 @@
+        (self.context == id.scope &&
+         self.flag_match(chain![prefix(negate), id.name_components_strs()])) ||
+        // Check if -s matches for a short name s, if any.
+        (if let Some(sn) = &id.short_name {
+            self.flag_match(chain![once(""), once(sn.as_ref())])
+        } else { false })
     }
 
-    pub fn argv() -> Self {
-        Self::new(env::args().collect::<Vec<_>>())
+    fn matches(&self, id: &OptionId) -> bool {
+        self._matches(id, false)
     }
 
-    fn arg_name(id: &OptionId, negate: Negate) -> String {
-        format!(
-            "--{}{}{}",
-            match negate {
-                Negate::False => "",
-                Negate::True => "no-",
-            },
-            match &id.scope {
-                Scope::Global => "".to_string(),
-                Scope::Scope(scope) => format!("{}-", scope.to_ascii_lowercase()),
-            },
-            id.name("-", NameTransform::ToLower)
-        )
+    fn matches_negation(&self, id: &OptionId) -> bool {
+        self._matches(id, true)
     }
 
-    fn arg_names(id: &OptionId, negate: Negate) -> HashMap<String, bool> {
-        let mut arg_names = HashMap::new();
-        if let Some(short_name) = &id.short_name {
-            arg_names.insert(format!("-{short_name}"), false);
-            if negate == Negate::True {
-                arg_names.insert(format!("--no-{short_name}"), true);
-            }
-        }
-        arg_names.insert(Self::arg_name(id, Negate::False), false);
-        if negate == Negate::True {
-            arg_names.insert(Self::arg_name(id, Negate::True), true);
+    fn to_bool(&self) -> Result<Option<bool>, ParseError> {
+        // An arg can represent a bool either by having an explicit value parseable as a bool,
+        // or by having no value (in which case it represents true).
+        match &self.value {
+            Some(value) => match expand(value.to_string())? {
+                Some(s) => bool::parse(&s).map(Some),
+                _ => Ok(None),
+            },
+            None => Ok(Some(true)),
         }
-        arg_names
     }
+}
 
-    fn find_flag(
-        &self,
-        flag_names: HashMap<String, bool>,
-    ) -> Result<Option<(String, String, bool)>, String> {
-        for arg in self.args.iter().rev() {
-            let mut components = arg.as_str().splitn(2, '=');
-            if let Some(name) = components.next() {
-                if let Some(negated) = flag_names.get(name) {
-                    return Ok(Some((
-                        name.to_owned(),
-                        components.next().unwrap_or("").to_owned(),
-                        *negated,
-                    )));
+pub struct Args {
+    args: Vec<Arg>,
+}
+
+impl Args {
+    pub fn new(arg_strs: Vec<String>) -> Self {
+        let mut args: Vec<Arg> = vec![];
+        let mut scope = Scope::Global;
+        for arg_str in arg_strs.into_iter() {
```

Comment:
Not super important, but the lack of `else` in the `if`-chain caught my eye as "potentially unintentional".... I've convinced myself that it is intentional/correct, but it might be nice to save the next person that confusion?

AIUI, this loop is filtering down to args that look like either `--`/`-` flags, or goals, i.e. just the args that can supply options, eliminating paths/targets & passthroughs.

So maybe we could either just add a comment or two, or add the `else` with a comment there `else { // a file path or target, not relevant for parsing options }`.

Another phrasing of the last version would be something like: